File: src/music/Rhythm.java
package music;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;


// Represents a musical rhythm as a list of note durations
// TODO: finish implementation in other SoundPlayer and other classes
public class Rhythm {
	private int bpm; // beats per minute
	private int beatDuration; // in milliseconds
	private ArrayList<Integer> durations; // duration of each note
	
	public Rhythm(int bpm) {
		this.bpm = bpm;
		this.beatDuration = (1000 * 60) / this.bpm;
	}

	public Rhythm() {
		this(120);
	}

	public void addNote(RhythmNote note) {
		durations.add(beatDuration * note.getBeats());
	}

	public void setBPM(int bpm) {
		this.bpm = bpm;
		this.beatDuration = (1000 * 60) / this.bpm;
	}

	static enum RhythmNote {
		Quarter, Half, Whole;

		public static final List<RhythmNote> notes = Arrays.asList(Quarter, Half, Whole);
		public static final List<Integer> beats = Arrays.asList(1, 2, 4);

		public int getBeats() {
			return beats.get(notes.indexOf(this));
		}
	}
}

File: src/music/Rhythm.py
from enum import Enum
from typing import List

class Rhythm:
    class RhythmNote(Enum):
        Quarter = 1
        Half = 2
        Whole = 4

        @classmethod
        def get_beats(cls, note: 'RhythmNote') -> int:
            return note.value

    def __init__(self, bpm: int = 120):
        self.bpm = bpm
        self.beat_duration = (1000 * 60) // self.bpm
        self.durations: List[int] = []

    def add_note(self, note: RhythmNote):
        self.durations.append(self.beat_duration * note.value)

    def set_bpm(self, bpm: int):
        self.bpm = bpm
        self.beat_duration = (1000 * 60) // self.bpm

# Small tests
if __name__ == "__main__":
    rhythm = Rhythm()
    assert rhythm.bpm == 120
    assert rhythm.beat_duration == 500

    rhythm.add_note(Rhythm.RhythmNote.Quarter)
    rhythm.add_note(Rhythm.RhythmNote.Half)
    rhythm.add_note(Rhythm.RhythmNote.Whole)
    assert rhythm.durations == [500, 1000, 2000]

    rhythm.set_bpm(60)
    assert rhythm.beat_duration == 1000

    print("All tests passed!")
